### ISSUE TITLE
TS2420 and TS2416 errors since @types/node 20.16.12

### ISSUE THREAD

**Issue by yoursunny** [zen-fs/core#147]
\@zenfs/core 1.4.2 is incompatible with \@types/node 20.16.12 or newer.
TypeScript compiler raises TS2420 and TS2416 errors.

## package.json

```json
{
  "private": true,
  "type": "module",
  "packageManager": "pnpm@9.14.2+sha512.6e2baf77d06b9362294152c851c4f278ede37ab1eba3a55fda317a4a17b209f4dbb973fb250a77abc463a341fcb1f17f17cfa24091c4eb319cda0d9b84278387",
  "dependencies": {
    "@types/node": "20.16.12",
    "@zenfs/core": "^1.4.2",
    "typescript": "~5.6.3"
  }
}
```

Same error also occurs with \@types/node 20.17.8 (latest 20.x version) and 22.10.0 (lastest 22.x version).

## tsconfig.json

```jsonc
{
  "compilerOptions": {
    "forceConsistentCasingInFileNames": true,
    "module": "ES2022",
    "moduleResolution": "Bundler",
    "noEmitOnError": true,
    "noImplicitOverride": true,
    "noImplicitReturns": true,
    "noUncheckedIndexedAccess": true,
    "noUnusedLocals": true,
    "strict": true,
    "target": "ESNext",
    "types": [
      "node"
    ]
  },
  "include": [
    "main.ts"
  ],
  "exclude": [
    "**/node_modules"
  ]
}
```

## main.ts

```ts
import { fs } from "@zenfs/core";

console.log(typeof fs.read);
```

## Command Line

```bash
./node_modules/.bin/tsc --noEmit
```

## Error Message

```text
node_modules/.pnpm/@types+readable-stream@4.0.18/node_modules/@types/readable-stream/index.d.ts:419:11 - error TS2420: Class 'Readable' incorrectly implements interface 'ReadableStream'.
  Types of property 'wrap' are incompatible.
    Type '(oldStream: Readable) => this' is not assignable to type '(oldStream: ReadableStream) => this'.
      Types of parameters 'oldStream' and 'oldStream' are incompatible.
        Type 'ReadableStream' is not assignable to type 'Readable'.
          Type 'ReadableStream' is missing the following properties from type 'Readable': readableAborted, readableDidRead, readableEncoding, readableEnded, and 28 more.

419     class Readable extends _Readable implements NodeJS.ReadableStream {
              ~~~~~~~~

node_modules/.pnpm/@zenfs+core@1.4.2/node_modules/@zenfs/core/dist/emulation/dir.d.ts:58:5 - error TS2416: Property '[Symbol.asyncIterator]' in type 'Dir' is not assignable to the same property in base type 'Dir'.
  Type '() => AsyncIterableIterator<Dirent>' is not assignable to type '() => AsyncIterator<Dirent, any, any>'.
    Property '[Symbol.asyncDispose]' is missing in type 'AsyncIterableIterator<Dirent>' but required in type 'AsyncIterator<Dirent, any, any>'.

58     [Symbol.asyncIterator](): AsyncIterableIterator<Dirent>;
       ~~~~~~~~~~~~~~~~~~~~~~

  node_modules/.pnpm/typescript@5.6.3/node_modules/typescript/lib/lib.esnext.disposable.d.ts:40:5
    40     [Symbol.asyncDispose](): PromiseLike<void>;
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    '[Symbol.asyncDispose]' is declared here.

node_modules/.pnpm/@zenfs+core@1.4.2/node_modules/@zenfs/core/dist/emulation/promises.d.ts:161:5 - error TS2416: Property 'createReadStream' in type 'FileHandle' is not assignable to the same property in base type 'FileHandle'.
  Type '(options?: import("fs/promises").CreateReadStreamOptions | undefined) => import("/home/ubuntu/zenfs-demo/node_modules/.pnpm/@zenfs+core@1.4.2/node_modules/@zenfs/core/dist/emulation/streams").ReadStream' is not assignable to type '(options?: import("fs/promises").CreateReadStreamOptions | undefined) => import("fs").ReadStream'.
    Call signature return types 'ReadStream' and 'ReadStream' are incompatible.
      The types of 'addListener' are incompatible between these types.
        Type '{ (event: "close", listener: () => void): ReadStream; (event: "data", listener: (chunk: any) => void): ReadStream; (event: "end", listener: () => void): ReadStream; (event: "readable", listener: () => void): ReadStream; (event: "error", listener: (err: Error) => void): ReadStream; (event: string | symbol, listener: ...' is not assignable to type '{ (event: "close", listener: () => void): ReadStream; (event: "data", listener: (chunk: string | Buffer) => void): ReadStream; (event: "end", listener: () => void): ReadStream; (event: "error", listener: (err: Error) => void): ReadStream; (event: "open", listener: (fd: number) => void): ReadStream; (event: "pause", ...'.
          Types of parameters 'event' and 'event' are incompatible.
            Type '"data"' is not assignable to type '"close"'.

161     createReadStream(options?: promises.CreateReadStreamOptions): ReadStream;
        ~~~~~~~~~~~~~~~~

node_modules/.pnpm/@zenfs+core@1.4.2/node_modules/@zenfs/core/dist/emulation/streams.d.ts:4:22 - error TS2720: Class 'import("/home/ubuntu/zenfs-demo/node_modules/.pnpm/@zenfs+core@1.4.2/node_modules/@zenfs/core/dist/emulation/streams").ReadStream' incorrectly implements class 'import("fs").ReadStream'. Did you mean to extend 'import("fs").ReadStream' and inherit its members as a subclass?
  Types of property 'wrap' are incompatible.
    Type '(oldStream: Readable) => this' is not assignable to type '(stream: ReadableStream) => this'.
      Types of parameters 'oldStream' and 'stream' are incompatible.
        Type 'ReadableStream' is not assignable to type 'Readable'.
          Type 'ReadableStream' is missing the following properties from type 'Readable': readableAborted, readableDidRead, readableEncoding, readableEnded, and 28 more.

4 export declare class ReadStream extends Readable implements Node.ReadStream {
                       ~~~~~~~~~~


Found 4 errors in 4 files.

Errors  Files
     1  node_modules/.pnpm/@types+readable-stream@4.0.18/node_modules/@types/readable-stream/index.d.ts:419
     1  node_modules/.pnpm/@zenfs+core@1.4.2/node_modules/@zenfs/core/dist/emulation/dir.d.ts:58
     1  node_modules/.pnpm/@zenfs+core@1.4.2/node_modules/@zenfs/core/dist/emulation/promises.d.ts:161
     1  node_modules/.pnpm/@zenfs+core@1.4.2/node_modules/@zenfs/core/dist/emulation/streams.d.ts:4
```

**Comment by james-pre** [comment]
2 Separate issues here: `Readable` stream not being compatible and one a missing property on `Dir[Symbol.asyncIterator]()`. The first issue might be an upstream issue with `readable-stream`, and is a little complicate. The latter issue I'll take a look at before the first one, which I should be able to fix with relative ease.

**Comment by james-pre** [comment]
Fixed in [v1.5.1](https://github.com/zen-fs/core/releases/tag/v1.5.1).

**Comment by yoursunny** [comment]
TS2420 is gone.
TS2416 is still there, re-posted as #153.
